src/App.ts

```diff
@@ -488,8 +490,10 @@ export default class App {
    *
    * @param m global middleware function
    */
-  public use(m: Middleware<AnyMiddlewareArgs>): this {
-    this.middleware.push(m);
+  public use<MiddlewareCustomContext extends StringIndexed = StringIndexed>(
+    m: Middleware<AnyMiddlewareArgs, AppCustomContext & MiddlewareCustomContext>,
+  ): this {
+    this.middleware.push(m as Middleware<AnyMiddlewareArgs>);
```

Comment:
Should the private `middleware` property also be typed as `Middleware<AnyMiddlewareArgs, AppCustomContext & MiddlewareCustomContext>[]`?

Reply:
Currently if it's typed with the Contexts the below compiler error is generated:
```
TS2345: Argument of type 'Middleware<AnyMiddlewareArgs, AppCustomContext & MiddlewareCustomContext>' is not assignable to parameter of type 'Middleware<AnyMiddlewareArgs, StringIndexed>'.
   Type 'StringIndexed' is not assignable to type 'AppCustomContext & MiddlewareCustomContext'.
     Type 'StringIndexed' is not assignable to type 'AppCustomContext'.
       'StringIndexed' is assignable to the constraint of type 'AppCustomContext', but 'AppCustomContext' could be instantiated with a different subtype of constraint 'StringIndexed'.
```

In addition, I'm not sure the typing is as important on this line, as once the middlewares are in the private global middlewares, I don't think Bolt cares about the custom context type anymore? From my limited testing, if
(Line: 228)
```ts
  /** Global middleware chain */
  private middleware: Middleware<AnyMiddlewareArgs>[];
```
is changed to 
```ts
  /** Global middleware chain */
  private middleware: Middleware<AnyMiddlewareArgs, AppCustomContext>[];
```

Then `m as Middleware<AnyMiddlewareArgs, AppCustomContext>` will work. But doing this, causes the `processMiddleware` function(https://github.com/slackapi/bolt-js/blob/main/src/middleware/process.ts) to want to be aware of the app context as well. Circling back to the question of whether Bolt cares about the context internally, I don't really see a scenario where the context will matter at that point as the consumer of the library only interacts with the middlewares via the `use`, `shortcut`, `message`, etc middleware registration points.

Reply:
Thank you for the explanation! I appreciate it; my TS isn't great so this is helpful 🙇